core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -0,0 +1,356 @@
+          producer.abortTransaction()
+          debug(s"aborted transaction Last committed record: ${new String(records.last.value(), "UTF-8")}. Num " +
+            s"records written to $topic2: $recordsProcessed")
+          consumer.close()
+          consumer = transactionalConsumer(consumerGroupId, maxPollRecords = numSeedMessages / 4)
+          consumer.subscribe(List(topic1))
+        }
+      }
+    } finally {
+      producer.close()
+      consumer.close()
+    }
+
+    // Inspite of random aborts, we should still have exactly 1000 messages in topic2. Ie. we should not
+    // re-copy or miss any messages from topic1, since the consumed offsets were committed transactionally.
+    val verifyingConsumer = transactionalConsumer("foobargroup")
+    verifyingConsumer.subscribe(List(topic2))
+    val valueSeq = TestUtils.pollUntilAtLeastNumRecords(verifyingConsumer, numSeedMessages).map { record =>
+      TestUtils.assertCommittedAndGetValue(record).toInt
+    }
+    verifyingConsumer.close()
+    val valueSet = valueSeq.toSet
+    assertEquals(s"Expected $numSeedMessages values in $topic2.", numSeedMessages, valueSeq.size)
+    assertEquals(s"Expected ${valueSeq.size} unique messages in $topic2.", valueSeq.size, valueSet.size)
+  }
+
+  @Test
+  def testFencingOnCommit() = {
+    val transactionalId = "my-t.id"
+    val producer1 = TestUtils.createTransactionalProducer(transactionalId, servers)
+    val producer2 = TestUtils.createTransactionalProducer(transactionalId, servers)
+    val consumer = transactionalConsumer()
+    consumer.subscribe(List(topic1, topic2))
+
+    try {
+      producer1.initTransactions()
+
+      producer1.beginTransaction()
+      producer1.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, "1", "1", willBeCommitted = false))
+      producer1.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic2, "3", "3", willBeCommitted = false))
+
+      producer2.initTransactions()  // ok, will abort the open transaction.
+      producer2.beginTransaction()
+      producer2.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, "2", "4", willBeCommitted = true))
+      producer2.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic2, "2", "4", willBeCommitted = true))
+
+      try {
+        producer1.commitTransaction()
+        fail("Should not be able to commit transactions from a fenced producer.")
+      } catch {
+        case e : ProducerFencedException =>
+          // good!
+          producer1.close()
+        case e : Exception =>
+          fail("Got an unexpected exception from a fenced producer.", e)
+      }
+
+      producer2.commitTransaction()  // ok
+
+      val records = pollUntilExactlyNumRecords(consumer, 2)
+      records.zipWithIndex.foreach { case (record, i) =>
+        TestUtils.assertCommittedAndGetValue(record)
+      }
+    } finally {
+      consumer.close()
+      producer2.close()
+    }
+  }
+
+  @Ignore @Test
+  def testFencingOnSend() {
```

Comment:
Another case is fencing on `sendOffsets`. I'll leave it up to you whether it's worthwhile or not.